### ISSUE TITLE
Customer Address Custom Attribute gets deleted upon update

### ISSUE THREAD

**Issue by marcoreni** [magento/magento2#6411]
### Preconditions

1. Magento 2.1.0 CE, php 7.0.8
### Steps to reproduce
1. Define new Customer Address Custom Attribute in UpgradeData:

``` php
$customerSetup = $this->customerSetupFactory->create(['setup' => $setup]);
$addressEntity = $customerSetup->getEavConfig()->getEntityType('customer_address');
$attributeSetId = $addressEntity->getDefaultAttributeSetId();

/** @var AttributeSet $attributeSet  */
$attributeSet = $this->attributeSetFactory->create();
$attributeGroupId = $attributeSet->getDefaultGroupId($attributeSetId);

$customerSetup->addAttribute('customer_address','attribute_test',[
                'type'          => 'varchar',
                'label'         => 'Attribute Test',
                'input'         => 'text',
                'required'      => false,

                'visible'       => true,
                'user_defined'  => false,
                'position'      => 200,
                'sort_order'    => 200,
                'system'        => false,
            ]);

$attr1 = $customerSetup->getEavConfig()->getAttribute('customer_address', 'attribute_test');
$attr1->addData($customerAddressData = [
                'attribute_set_id' => $attributeSetId,
                'attribute_group_id' => $attributeGroupId,
                'used_in_forms' => ['adminhtml_customer_address']);
$this->attributeRepository->save($attr1);
```
1. Populate the attribute through adminhtml interface. The attribute is correctly persisted on EAV tables
2. Intercept the LoginPost action with a beforeExecute(), then retrieve and save customer without modifications:

``` php
$customer = $this->customerRepositoryInterface->get($login['username']);
$this->customerRepositoryInterface->save($customer);
return;
```
### Expected result

1. The Custom Attribute should stay there.
### Actual result

1. The Custom Attribute gets deleted

It seems that the problem may reside in Magento\Customer\Model\ResourceModel\Customer\Relation.php:49 which uses $address->save() instead of the repository (thus not having an attribute_set_id set), but if I remove the ->save($customer)  in the plugin everything works as expected.

**Comment by marcoreni** [comment]
Any update about this?

**Comment by rzinnatullin** [comment]
Seems like the same issue here. But... are you sure you are talking about 'Custom Attribute'? In my case, Custom Attribute doesn't get affected, but Custom Attribute **Value** is deleted from customer_address_entity_varchar table.
Magento 2.1.2

**Comment by rzinnatullin** [comment]
In my case the reproduction steps are the same as @marcoreni described, but instead of resaving a customer I do this:

``` php
    $addr = $this->addressFactory->create()->load($addressId);
    $addr->save();
```

To workaround the issue I put `$addr->setAttributeSetId(2)` before `$addr->save()`

**Comment by sevos1984** [comment]
The problem is in attributeRepository, as workaround you can use $attr1->save() instead $this->attributeRepository->save($attr1); This is a Magento issue, I've created ticket for that MAGETWO-60815. Thanks for reporting

**Comment by andreaschiffo** [comment]
Hi
i found the same issue: added a custom customer address field and when i insert a new address, it gets inserted in  customer_address_entity_varchar but when i update the address, the value is deleted

may this be due to /vendor/magento/module-customer/Model/Address/CustomAttributeList.php which returns an empty array?
....that nobody's gonna populate...
this happens only on update because when you insert the address, Magento doesn't check whether the attribute is among the custom ones....but during update it does check the presence of your attribute among the valid custom attributes...the problem is in fact that the valid attributes is an empty array

e.g.

**shop/vendor/magento/module-customer/Model/ResourceModel/AddressRepository.php** at line 114 it does
`$addressModel->updateData($address);`
because the addressModel is not null (Magento has found an existing address, therefore it wont insert but update)

**shop/vendor/magento/module-customer/Model/Address.php** at line 149 executes the magic method setCustomAttributes
`$this->setDataUsingMethod($attributeCode, $attributeData);`

**then it goes at shop/vendor/magento/framework/Model/AbstractExtensibleModel.php** line 157, where it does setData, which filters custom attributes: **shop/vendor/magento/framework/Model/AbstractExtensibleModel.php** line 80

in order to filter them it collects the customAttributeCodes

`$customAttributesCodes = $this->getCustomAttributesCodes();`

and this simply returns an empty array, therefore there wont ever be any valid custom attributes for customer address

**shop/vendor/magento/framework/Model/AbstractExtensibleModel.php** line 275

```
 protected function getCustomAttributesCodes()
    {
        return [];
    }
```

im currently using 2.1.0

**Comment by rzinnatullin** [comment]
The issue happens even if you update customer's billing address from account page.
Magento 2.1.2
I have an extension which subscribed to customer_address_after_save event. This is the call stack:
```
<custom observer method here>
Event/Invoker/InvokerDefault.php.Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod:73	
Event/Invoker/InvokerDefault.php.Magento\Framework\Event\Invoker\InvokerDefault->dispatch:61	
Event/Manager.php.Magento\Framework\Event\Manager->dispatch:66	
var/generation/Magento/Framework/Event/Manager/Proxy.php.Magento\Framework\Event\Manager\Proxy->dispatch:95	
Model/AbstractModel.php.Magento\Framework\Model\AbstractModel->afterSave:803	
Model/Address.php.Magento\Customer\Model\Address->afterSave:330	
Model/Entity/VersionControl/AbstractEntity.php.Magento\Eav\Model\Entity\VersionControl\AbstractEntity->save:94	
Interception/Interceptor.php.Magento\Customer\Model\ResourceModel\Address\Interceptor->___callParent:74	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:70	
Interception/Interceptor.php.Magento\Customer\Model\ResourceModel\Address\Interceptor->Magento\Framework\Interception\{closure}:138	
App/Cache/FlushCacheByTags.php.Magento\Framework\App\Cache\FlushCacheByTags->aroundSave:60	
Interception/Interceptor.php.Magento\Customer\Model\ResourceModel\Address\Interceptor->___callPlugins:142	
var/generation/Magento/Customer/Model/ResourceModel/Address/Interceptor.php.Magento\Customer\Model\ResourceModel\Address\Interceptor->save:39	
Model/AbstractModel.php.Magento\Framework\Model\AbstractModel->save:632	
Model/ResourceModel/AddressRepository.php.Magento\Customer\Model\ResourceModel\AddressRepository->save:121	
Controller/Address/FormPost.php.Magento\Customer\Controller\Address\FormPost->execute:193	
App/Action/Action.php.Magento\Framework\App\Action\Action->dispatch:102	
Controller/Address.php.Magento\Customer\Controller\Address->dispatch:128	
Interception/Interceptor.php.Magento\Customer\Controller\Address\FormPost\Interceptor->___callParent:74	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:70	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
App/Action/Plugin/Context.php.Magento\Store\App\Action\Plugin\Context->aroundDispatch:106	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
Model/App/Action/ContextPlugin.php.Magento\Customer\Model\App\Action\ContextPlugin->aroundDispatch:61	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
App/Action/Plugin/StoreCheck.php.Magento\Store\App\Action\Plugin\StoreCheck->aroundDispatch:44	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
Model/App/Action/ContextPlugin.php.Magento\Weee\Model\App\Action\ContextPlugin->aroundDispatch:112	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Interceptor.php.Magento\Customer\Controller\Address\FormPost\Interceptor->Magento\Framework\Interception\{closure}:138	
Model/App/Action/ContextPlugin.php.Magento\Tax\Model\App\Action\ContextPlugin->aroundDispatch:91	
Interception/Interceptor.php.Magento\Customer\Controller\Address\FormPost\Interceptor->___callPlugins:142	
file:///var/www/html/magento/var/generation/Magento/Customer/Controller/Address/FormPost/Interceptor.php.Magento\Customer\Controller\Address\FormPost\Interceptor->dispatch:26	
App/FrontController.php.Magento\Framework\App\FrontController->dispatch:55	
Interception/Interceptor.php.Magento\Framework\App\FrontController\Interceptor->___callParent:74	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:70	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
App/FrontController/Plugin/RequestPreprocessor.php.Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch:94	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
Module/Plugin/DbStatusValidator.php.Magento\Framework\Module\Plugin\DbStatusValidator->aroundDispatch:69	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}:63	
Model/App/FrontController/VarnishPlugin.php.Magento\PageCache\Model\App\FrontController\VarnishPlugin->aroundDispatch:55	
Interception/Chain/Chain.php.Magento\Framework\Interception\Chain\Chain->invokeNext:67	
Interception/Interceptor.php.Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}:138	
Model/App/FrontController/BuiltinPlugin.php.Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch:68	
Interception/Interceptor.php.Magento\Framework\App\FrontController\Interceptor->___callPlugins:142	
var/generation/Magento/Framework/App/FrontController/Interceptor.php.Magento\Framework\App\FrontController\Interceptor->dispatch:26	
App/Http.php.Magento\Framework\App\Http->launch:135	
App/Bootstrap.php.Magento\Framework\App\Bootstrap->run:258	
index.php.{main}:39
```
And this is how we lose custom attributes values (the screenshot of debugger's Variables tab, breakpoint is inside customer_address_after_save event handler):
![image](https://cloud.githubusercontent.com/assets/11244866/25285237/00376d32-26d3-11e7-881e-dca3b3f40fa2.png)

I belive the issue is a showstopper since it's impossible to use custom attributes in customer address model.
Any info when the issue is going to be fixed?

**Comment by rzinnatullin** [comment]
Seems like the issue I described above has another nature.
We click 'Save Address' in a customer's Address Book. What then happens inside Magento 2:
1. FormPost.php/execute
```` php
$address = $this->_extractAddress();
````
2. FormPost.php/_extractAddress
```` php
$addressData = $addressForm->extractData($this->getRequest())
````
3. Form.php/extractData
```` php
        $data = [];
        foreach ($this->getAllowedAttributes() as $attribute) {
            $dataModel = $this->_getAttributeDataModel($attribute);
            $dataModel->setRequestScope($scope);
            $dataModel->setRequestScopeOnly($scopeOnly);
            $data[$attribute->getAttributeCode()] = $dataModel->extractValue($request);
        }
        return $data;
````
4. Text.php/extractValue
``` php
return $this->_applyInputFilter($this->_getRequestValue($request));
```
5. AbstractData.php/_getRequestValue
``` php
        } else {
            $value = $request->getParam($attrCode, false);
        }
```
6.  Request.php/getParam
``` php
        $key = (string) $key;
        $keyName = (null !== ($alias = $this->getAlias($key))) ? $alias : $key;
        if (isset($this->params[$keyName])) {
            return $this->params[$keyName];
        } elseif (isset($this->queryParams[$keyName])) {
            return $this->queryParams[$keyName];
        } elseif (isset($this->postParams[$keyName])) {
            return $this->postParams[$keyName];
        }
        return $default;
```
For our custom attribute it cannot find a value (params, queryParams and postParams don't contain a key for our custom attribute), so it returns $default which is = FALSE.
Let's get back to FormPost.php/execute:
``` php
            $address = $this->_extractAddress();
            $this->_addressRepository->save($address);
```
Have a look what $address->_data contains:
![image](https://cloud.githubusercontent.com/assets/11244866/25288826/84ca46c0-26e0-11e7-9881-4e3e70596347.png)
It contains custom attributes with empty values, because $request didn't contain them in the step 6 (above).
In AddressRepository.php/save we can see
``` php
            $addressModel->updateData($address);
```
In Address.php/updateData:
``` php
        // Set all attributes
        $attributes = $this->dataProcessor
            ->buildOutputDataArray($address, '\Magento\Customer\Api\Data\AddressInterface');

        foreach ($attributes as $attributeCode => $attributeData) {
            if (AddressInterface::REGION === $attributeCode) {
                $this->setRegion($address->getRegion()->getRegion());
                $this->setRegionCode($address->getRegion()->getRegionCode());
                $this->setRegionId($address->getRegion()->getRegionId());
            } else {
                $this->setDataUsingMethod($attributeCode, $attributeData);
            }
        }
```
buildOutputDataArray gives attributes list with empty custom attributes:
![image](https://cloud.githubusercontent.com/assets/11244866/25288592/71974c7a-26df-11e7-9284-0b5131f6a9d5.png)
And then it overwrites existing custom attributes values with empty ones:
``` php
            } else {
                $this->setDataUsingMethod($attributeCode, $attributeData);
            }
```
Is it a bug or by design?
In my opinion, if an attribute's value wasn't received from the form then it should be left unchanged (not overwritten with some default value).

**Comment by TheFrankman** [comment]
As a work around for this I have created a plugin. This plugin adds my custom attribute to the list of allowed attributes when the admin for saves. 

Vendor/Namespace/etc/di.xml

```
 <!-- Plugin to ensure custom address attributes get saved in admin -->
    <type name="Magento\Customer\Model\Metadata\AddressMetadata">
        <plugin name="saveCustomAddressAttributes" type="Vendor\Namespace\Plugin\SaveCustomAddressAttributes" sortOrder="0" disabled="false"/>
    </type>
```
Vendor/Namespace/Plugin/SaveCustomAddressAttributes.php

```
<?php

namespace Vendor\Namespace\Plugin;

use Magento\Customer\Api\AddressMetadataInterface;
use Magento\Customer\Model\AttributeMetadataConverter;
use Magento\Customer\Model\AttributeMetadataDataProvider;

/**
 * Class SaveCustomAddressAttributes
 * @package Vendor\Namespace\Plugin
 * @author Frank Clark
 */
class SaveCustomAddressAttributes
{
    /** @var AttributeMetadataConverter  */
    protected $attributeMetadataConverter;
    /** @var AttributeMetadataDataProvider  */
    protected $attributeMetadataDataProvider;
    /** @var array  */
    protected $additionalAddressAttributes = array(
        'pic_id'
    );

    /**
     * SaveCustomAddressAttributes constructor.
     * @param AttributeMetadataConverter $attributeMetadataConverter
     * @param AttributeMetadataDataProvider $attributeMetadataDataProvider
     */
    public function __construct(
        AttributeMetadataConverter $attributeMetadataConverter,
        AttributeMetadataDataProvider $attributeMetadataDataProvider
    ) {
        $this->attributeMetadataConverter = $attributeMetadataConverter;
        $this->attributeMetadataDataProvider = $attributeMetadataDataProvider;
    }

    /**
     * Work around to ensure that custom address attribute get saved
     * @param \Magento\Customer\Model\Metadata\AddressMetadata $subject
     * @param array $attributes
     * @return array
     */
    public function afterGetAttributes(\Magento\Customer\Model\Metadata\AddressMetadata $subject, $attributes)
    {
        // Loop through our custom address attributes
        foreach ($this->additionalAddressAttributes as $additionalAddressAttribute) {
            // Only add attribute if ti hasn't already been added
            if (!isset($attributes[$additionalAddressAttribute])) {
                // Create an instance of our attribute
                try {
                    $attribute = $this->attributeMetadataDataProvider
                        ->getAttribute(
                            AddressMetadataInterface::ENTITY_TYPE_ADDRESS,
                            $additionalAddressAttribute
                        );
                } catch (\Magento\Framework\Exception\LocalizedException $e) {
                    // If for whatever reason our attribute doesn't exist then return the original array
                    return $attributes;
                }

                if ($attribute) {
                    // If we have retrieved an attribute then get the meta and add it to the attributes array
                    $convertedAttribute = $this->attributeMetadataConverter->createMetadataAttribute($attribute);
                    if ($convertedAttribute) {
                        $attributes[$additionalAddressAttribute] = $convertedAttribute;
                    }
                }

            }
        }
        return $attributes;
    }
}

```

In theory because we are checking ` if (!isset($attributes[$additionalAddressAttribute])) {` if Magento resolve this issue later it wont cause a problem. 

Just add whatever your custom attributes are to $this->additionalAddressAttributes

**Comment by magento-engcom-team** [comment]
@marcoreni, thank you for your report.
We've created internal ticket(s) MAGETWO-60815 to track progress on the issue.

**Comment by Gatzzu** [comment]
#WroCD

**Comment by Gatzzu** [comment]
#WroCD

**Comment by ecomchallenger** [comment]
Unable to reproduce the original issue on branch: 2.2-develop (commint: 6f8fc455560f34a773a9a116d5548b60142afd59).

By the way during trying to reproduce the issue, I encountered another issue, which might be the reason of the customer address custom attribute value reset in case from issue description.

This another issue is that entries from database table: customer_form_attribute are not cleared on removing attribute via install / upgrade script, which might lead to situation, when attribute is added to more forms, than intended (when for example upgrade script was invoked multiple times).

The only way, in which I manage to get a similar result, to the original issue, was the case, when customer address custom attribute was set to use in forms both: "adminhtml_customer_address" and "customer_address_edit". In this configuration, value of custom attribute was properly saved via admin area, but on address edit in Customer address Book on frontend the value was reset. Why the field is not visible in customer address book for mentioned configuration - this might be another issue or feature request.

Regarding to checkout issue, when customer address custom attribute is not saved in database, please refer to the [documentation](https://devdocs.magento.com/guides/v2.0/howdoi/checkout/checkout_new_field.html) to see the solution.

**Comment by ksiamro** [comment]
I have similar issue but custom attributes from all addresses (related to customer) are lost upon delete one address at frontend.

**Comment by Thundar** [comment]
Is this issue possibly fixed by MAGETWO-69111 from Magento version 2.2.0? Can you still reproduce it in this version or above?
Commit: https://github.com/magento/magento2/commit/0690ed44400e435eccf25cafe3e25df7485610f0

**Comment by LiamKarlMitchell** [comment]
Customer address book template is hard coded right?
Not built dynamically using the Knockout template/code that's used in the checkout.

**Comment by krzksz** [comment]
This issue should be resolved by 0690ed44400e435eccf25cafe3e25df7485610f0. Feel free to reopen if you manage to reproduce it on `2.4-develop` instance.
